### ISSUE TITLE
Add ability to export to an external editor within the same folder as the original file

### ISSUE THREAD

**Issue by Stuntflyer** [Beep6581/RawTherapee#6195]
Within the current build of RT 5.8, the program does not know the location of the original raw file when it exports to an external editor. Instead, it auto exports to a TEMP folder. It would be convenient if there where a user option to automatically save exports to the location of the original RAW file. This option is already available within the Processing Queue with the %p1/%f command.

For users that always save to a different folder than the original, it might not be an issue. However, I think that many photographers would like to keep all of their edits in the same folder as the original RAW file. For those users having to click on SAVE AS, pick the desired drive letter and then pick the location, is just a lot of steps that could be avoided if this feature was added.

I suggest either something similar to what is available in the Processing Queue or a Preference option where a path is stated for the external editor. 

Mike

**Comment by AndyAstbury** [comment]
Saving the RT output image to the parent folder of the original raw file makes perfect sense - it's the way virtually every other bit of photo software works!
When using the 'open in external editor' command, the RT output would be better served by being dropped alongside the original source raw, rather than being placed in a TEMP folder which disappears, along with its content, when the machine is turned off.  This leads to all manner of confusion and frustration amongst users new to Raw Therapy.

So Mikes suggestion/request has my full and enthusiastic support.

**Comment by TechXavAL** [comment]
Even though I like the idea (I wouldn't use it), I think there are some aspects to consider, like in which format the raw file will be exported to (automatically).

Let me explain a situation that I would face when using this feature: after processing my raw file I would like to open it in my preferred editor, but I would always sent the exported image as TIFF, to preserve quality as much as I can. Then after editing the image and going back to RT, maybe I would make a final processing, just because I like that RT tool better. After that, I would export my final image as JPEG.

So what has to be considered is how to tell RT which formats to use on each case, and make it logical and understandable to the user. And if the final exported image (my JPEG) would have to be saved next to the raw file or in a subfolder (as currently by default).

It's not impossible, just has to be planned before starting to code.

Anyway I think it would be nice having such feature.

**Comment by Stuntflyer** [comment]
@TechXavAL , all we are suggesting is an option for saving the tif and jpg files created in RT to the original folder location. The exported file formats would remain the same as they are now. No change at all. So, there is no reason to be concerned about "how to tell RT which formats to use on each case, and make it logical and understandable to the user".

Saving an image to a destination that would be wiped out if you turn the machine off or worse yet a power cut or crash is not a good idea and the option to avoid such things should be available.

**Comment by TechXavAL** [comment]
We mostly agree.

But then I have a doubt: when you send to the external editor which format should be used? Tiff, jpeg, or both? Or the user chooses which one?

**Comment by Stuntflyer** [comment]
Think of it as being no different than the way it is now other than where the Tiff is being stored while it is being edited in an external editor. Instead of being stored in a temp folder where the file is subject to being lost due to computer failure e.g., brown outs, the Tiff file is stored in the original folder where the RAW was launched from in the RT File Browser. 

Jpg options remain the same as the default RT behavior. That is "Save the Current Image" or "Processing Queue" function.

I am only asking for an **_option_**, so no one has to use it unless they prefer to work this way.  But, why would anyone want to risk losing their work due to computer shut downs or forgetting to change the location of the Tiff edit before they turn off the computer.

**Comment by heckflosse** [comment]
> Think of it as being no different than the way it is now other than where the Tiff is being stored while it is being edited in an external editor.

There is one special thing when saving to the current folder. The current folder is scanned in background periodically for new files. If there is a new file in the current folder which is not completely written at the moment the current folder is scanned, that may lead to errors. If we export to current folder, we have to change some things in code to stop the scanning of the current folder for the time the exported file is written.

**Comment by heckflosse** [comment]
> But then I have a doubt: when you send to the external editor which format should be used? Tiff, jpeg, or both?

imho it should be uncompressed 16 bit tiff

**Comment by Stuntflyer** [comment]
I agree, 16 bit tiff.

**Comment by TechXavAL** [comment]
I think there's enough information now to clearly understand what is asked for and what is not.

And now that there has been more activity in this thread, there's a higher chance that a developer becomes interested in it and maybe it gets implemented. :wink:

FWIW, even though I believe it's a useful feature to have in RT, I said I wouldn't use it because whenever I use GIMP from RT, it's not uncommon that I develop at least a couple versions, so in the end I have to use «Save as...». So to me, this is a feature I could easily live without. I guess not so much for other people.

**Comment by Stuntflyer** [comment]
> we have to change some things in code to stop the scanning of the current folder for the time the exported file is written.

Ingo, would this be the only real issue that would need to be resolved?

**Comment by heckflosse** [comment]
@Stuntflyer 

> would this be the only real issue that would need to be resolved?

I think so.

**Comment by AndyAstbury** [comment]
Just so we are clear I made this quick video [https://youtu.be/2jxS-Ml_7Ns](https://youtu.be/2jxS-Ml_7Ns)
If I've left anything out Mike just say!

**Comment by Stuntflyer** [comment]
Excellent, Andy.

Now I know that some RT users say that they have no need for this feature and that's all well and good. However, as an option, no changes would take place for those users unless they choose it. Additionally, if the option is available in the Processing Queue %p1/%f, then it makes sense to have the option in the Export to External Editor as it maintains continuity in the workflow. 

Whether or not the proposed option becomes the default behavior, as I think it should, remains to be seen. If it was, then new users unaware of the TEMP holding place would be protected. 

A warning message that can be disabled would give the user awareness of the new default and the opportunity to change it in preferences.

**Comment by waynesutton50** [comment]
This feature seems to be in the latest ART build
![Screenshot from 2021-05-07 18-48-39](https://user-images.githubusercontent.com/73446985/117482973-6ec69d80-af65-11eb-9d1a-771ed4a863d7.png)

**Comment by Stuntflyer** [comment]
Aha! Seems like a perfect solution that would keep everyone happy. Could this be ported over to RT?

**Comment by Desmis** [comment]
@Stuntflyer @AndyAstbury @waynesutton50 @Thanatomanic @TechXavAL 

I also find for my personal use that having the file elsewhere than "temp", is better, especially "same as input image" .

I did a quick test to port some of the ART code, it seems to work, but I need to check.

If no one objects, and if no one has started work, especially Ingo @heckflosse , then I could propose a new branch to test later today.

jacques

**Comment by Thanatomanic** [comment]
@Desmis Thanks for looking at the ART implementation! If you can make a PR we can test to see if it works.

Side-note: I don't object at all to implement this, even include this is in 5.9 as long as it works. Though, if only some of the [274 older other proposed enhancements](https://github.com/Beep6581/RawTherapee/issues?q=is%3Aopen+is%3Aissue++label%3A%22type%3A+enhancement%22+) would get such attention as this one ;-)

**Comment by Desmis** [comment]
@Thanatomanic 
I check the code and the changes, and if everything works as I want, I will open a PR this morning  :)



Jacques

**Comment by Desmis** [comment]
I push a change in a new branch "output_folder"
https://github.com/Beep6581/RawTherapee/tree/output_folder

jacques

**Comment by heckflosse** [comment]
If we could move this
![grafik](https://user-images.githubusercontent.com/13733487/117565631-8495a900-b0b2-11eb-83fe-782cc8c08242.png)

a little but upwards...

**Comment by Desmis** [comment]
@heckflosse 
The text of "Output directory" is aligned with that of "Gimp instalation directory". 
But the fact that Alberto added a "frame" - and I don't see how we can do otherwise, - gives this impression of being lower

Modifying this seems to me (who is not a GUI specialist at all) difficult

**Comment by AndyAstbury** [comment]
For goodness sake leave it alone - it works just fine where it is.  If I can work it out then anyone can, so there is - imo - no need for any changes.

**Comment by Thanatomanic** [comment]
Please keep things civil. Everyone has the right to comment on whatever they see fit.

**Comment by AndyAstbury** [comment]
As Jacques said "Modifying this seems to me (who is not a GUI specialist at all) **difficult**

The fact that the facility is in there is of EPIC importance, its positioning isn't, so WHY make work when you don't have to???
As for me being UNCIVIL - well....................

**Comment by heckflosse** [comment]
As the misalignment is only in preferences dialog (which is opened not so often), the misalignment is not a reason to not merge this.

But this is something which needs to be fixed:

![grafik](https://user-images.githubusercontent.com/13733487/117573536-bf123c80-b0d8-11eb-9e25-c677caab6064.png)

I saved to the folder `Z:\Eigene Bilder`. Obviously the path (which includes a space) is not passed correctly (to GIMP in this case)

**Comment by heckflosse** [comment]
I just checked. It's not related to the change. But exporting a file which has a space in filename or foldername (in case of saving to current folder) fails to open in GIMP.
@AndyAstbury Can you test this case with photoshop please? Just rename a file so it has a space in its filename and try to export this file from RT to photoshop?

**Comment by Desmis** [comment]
@heckflosse 

Ingo
I just tested with a file - as "play raw.nef"...
No problem to open in preview, and export in current folder to Photoshop (my very old version) works 

Jacques

**Comment by heckflosse** [comment]
@Desmis Ok, maybe this happens only when exporting to GIMP.

Nevertheless I will try to fix this and also let's wait for a test from @AndyAstbury

**Comment by Desmis** [comment]
I  tried  - always with Photoshop - and "play raw.nef"
* in current folder
* in "os temp dir"
All works fine
But of course  let's wait for a test from Andy

I am on Windows8, and my images are in F:Cou
Jacques

**Comment by Stuntflyer** [comment]
Only because I'm not totally familiar with the process do I ask this. Can the feature be merged into an upcoming release or does this have to wait until 5.9?

Mike

**Comment by heckflosse** [comment]
> Can the feature be merged into an upcoming release or does this have to wait until 5.9?

5.9 is the upcoming release

**Comment by AndyAstbury** [comment]
I'm on MAC am I not?
No MAC build with this feature.
So I run virtual LINUX - and Adobe do not service Linux so I can't test it with Photoshop.

However, it works perfectly sending to GIMP. The image opens correctly and the file resides in the parent raw folder (this folder is shared between MacOS and Linux).
Loading the image into MacOS Photoshop 2021 reveals the image is my desire colour space and at my desired 300ppi resolution.

So it works perfectly in so far as I can test it.

I'm not paying Microsoft for a Win license!!

**Comment by Stuntflyer** [comment]
> 
> 
> > Can the feature be merged into an upcoming release or does this have to wait until 5.9?
> 
> 5.9 is the upcoming release

I should have asked if the feature could be implemented in an upcoming dev build?

**Comment by Floessie** [comment]
@heckflosse 
> But this is something which needs to be fixed:

Could it be #5463?

**Comment by Desmis** [comment]
@Floessie  @heckflosse 

If the "problem" is  " " (white space) in folder, the simple code here solved that (in Editorpanel.cc - idle_sendToGimp)

        while (Glib::file_test (fileName, Glib::FILE_TEST_EXISTS) && tries < 1000) {
            fileName = Glib::ustring::compose ("%1-%2.%3", fname, tries, sf.format);
            tries++;
        }
       
        //printf("Filename=%s \n", fileName.c_str());
        for(int i=0; i < (int) fileName.length(); i++) {
            std::size_t foundws = fileName.find(' ');
            std::string str2="_";//replace " " by "_"
            if (foundws!=std::string::npos){
              //  std::cout << "ws found at: " << foundws << '\n';
                fileName.replace(foundws, 1, str2);
                printf("Filename new=%s \n", fileName.c_str());
            }
        }
I tried with Photoshop (which does not need this patch to work) or with Gimp...

jacques

**Comment by Desmis** [comment]
I pushed a possible fix, if "white space" are in folder;
It's not related to the change, but it prevents Gimp from working. 
Now it works, I replace " " by "_"

Jacques

**Comment by Desmis** [comment]
I added "verbose" in the console when "white-space" are found in folder ("white spaces" are replaced by "_")

Under "windows 8", everything works correctly (on my system) when there are "white-spaces", both with Gimp and Photoshop (or the replacement is not mandatory)
Has anyone tried under other systems (Linux,...)

Jacques

**Comment by Lawrence37** [comment]
If I export to a folder with a space in it, this will cause the image to be saved in a different folder, which is not what a user expects. We should figure out why spaces do not work and fix that instead of overwriting spaces.

**Comment by Desmis** [comment]
OK, only if White space are in the "folder"...,  but this behavior is not reported to this issue and its old...! what we do 

jacques

**Comment by Lawrence37** [comment]
Just for information, it works fine on Linux. Exporting to GIMP using a folder with spaces works correctly even without converting spaces to underscores.

**Comment by Desmis** [comment]
That confirm what says Andy, It seems only a windows problem..with Gimp

Jacques

**Comment by Desmis** [comment]
Even if it's not the "good" solution, I change replace "WS" by "_" only if "windows" and if user choose "Gimp" as external editor...

**Comment by Thanatomanic** [comment]
I think the fix for Windows is as easy as this:
```diff
diff --git a/rtgui/editorpanel.cc b/rtgui/editorpanel.cc
index e1e7ae30f..4dd3249cb 100644
--- a/rtgui/editorpanel.cc
+++ b/rtgui/editorpanel.cc
@@ -2085,25 +2085,6 @@ bool EditorPanel::idle_sendToGimp ( ProgressConnector<rtengine::IImagefloat*> *p
             tries++;
         }
 
-#ifdef __WIN32__   //only for Windows
-        if(options.editorToSendTo == 1){//if user choose Gimp
-            bool foundwhitespace = true;
-            while (foundwhitespace) {
-                std::size_t foundws = fileName.find(' ');
-                std::string str2="_";//replace " " by "_" but we can replace with another
-                if (foundws!=std::string::npos){
-                    fileName.replace(foundws, 1, str2);
-                    if (options.rtSettings.verbose) {
-                        printf("New file name without white-space=%s \n", fileName.c_str());
-                    }
-                } else {
-                    foundwhitespace = false;
-                }
-            
-            }
-        }
-#endif
-
         if (tries == 1000) {
             delete img;
             return false;
diff --git a/rtgui/extprog.cc b/rtgui/extprog.cc
index 57d57ecd8..f3a10de8c 100644
--- a/rtgui/extprog.cc
+++ b/rtgui/extprog.cc
@@ -276,7 +276,8 @@ bool ExtProgStore::openInGimp (const Glib::ustring& fileName)
     for (auto ver = 12; ver >= 0; --ver) {
 
         executable = Glib::build_filename (options.gimpDir, "bin", Glib::ustring::compose (Glib::ustring("gimp-2.%1.exe"), ver));
-        auto lsuccess = ShellExecute( NULL, "open", executable.c_str(), fileName.c_str(), NULL, SW_SHOWNORMAL );
+        Glib::ustring escapedFileName = Glib::ustring::compose ("\"%1\"", fileName);
+        auto lsuccess = ShellExecute( NULL, "open", executable.c_str(), escapedFileName.c_str(), NULL, SW_SHOWNORMAL );
 
         if (reinterpret_cast<uintptr_t>(lsuccess) > 32) {
             return true;
```

Edit: this should do it, the white space solution is no longer needed

**Comment by Desmis** [comment]
@Lawrence37 
I test your patch, works fine.
No more problem with Windows and gimp
And for other cases, Linux, Photoshop...all works fine

I think we can say  the bug is fixed :)

I just push the commit 09ae3f6a1

Thank you

Jacques

**Comment by Desmis** [comment]
@Thanatoma @heckflosse @Lawrence37  @AndyAstbury  @waynesutton50  @TechXavAL @Stuntflyer @Floessie

Thank you all for your contribution :)

merge :)

jacques

**Comment by Lawrence37** [comment]
@Desmis The patch was made by @Thanatomanic 😉

**Comment by Desmis** [comment]
@Thanatomanic  @Lawrence37 

OK, thank you, excuse me for the confusion 

jacques

**Comment by Stuntflyer** [comment]
Thank you all for your contribution in making this an RT feature. 

@Stuntflyer Mike
